rust_icu_uloc/src/lib.rs

```diff
@@ -346,43 +496,95 @@ pub fn accept_language(
+        let ptr = versioned_function!(uloc_toUnicodeLocaleType)(
+            legacy_keyword.as_ptr(),
+            legacy_value.as_ptr(),
+        );
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    unicode_value.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+/// Implements `uloc_toLegacyKey` from ICU4C.
+pub fn to_legacy_key(unicode_keyword: &str) -> Option<String> {
+    let unicode_keyword = str_to_cstring(unicode_keyword).unwrap();
+    let legacy_keyword: Option<ffi::CString> = unsafe {
+        let ptr = versioned_function!(uloc_toLegacyKey)(unicode_keyword.as_ptr());
+        ptr.as_ref().map(|ptr| ffi::CStr::from_ptr(ptr).to_owned())
+    };
+    legacy_keyword.map(|cstring| cstring_to_string(&cstring).unwrap())
+}
+
+fn str_to_cstring(input: &str) -> Result<ffi::CString, common::Error> {
+    ffi::CString::new(input).map_err(|_| common::Error::string_with_interior_nul())
+}
+
+fn cstring_to_string(input: &ffi::CString) -> Result<String, common::Error> {
+    input
+        .to_str()
+        .map(|s| s.to_owned())
+        .map_err(|e| common::Error::Wrapper(e.to_string().into()))
+}
+
 #[cfg(test)]
 mod tests {
-    use super::*;
+    use {super::*, anyhow::Error};
+
+    #[test]
+    fn test_language() -> Result<(), Error> {
+        let loc = ULoc::try_from("es-CO")?;
+        assert_eq!(loc.language(), Some("es".to_string()));
+        Ok(())
+    }
 
     #[test]
-    fn test_language() {
-        let loc = ULoc::try_from("es-CO").expect("get es_CO locale");
-        let language = loc.language().expect("should get language");
-        assert_eq!(&language, "es");
+    fn test_language_absent() -> Result<(), Error> {
+        let loc = ULoc::for_language_tag("und-CO")?;
+        assert_eq!(loc.language(), None);
+        Ok(())
     }
 
     #[test]
-    fn test_script() {
-        let loc = ULoc::try_from("sr-Cyrl").expect("get sr_Cyrl locale");
-        let script = loc.script().expect("should get script");
-        assert_eq!(&script, "Cyrl");
+    fn test_script() -> Result<(), Error> {
+        let loc = ULoc::try_from("sr-Cyrl")?;
+        assert_eq!(loc.script(), Some("Cyrl".to_string()));
+        Ok(())
     }
 
     #[test]
-    fn test_script_absent() {
-        let loc = ULoc::try_from("sr").expect("get sr locale");
-        let script = loc.script().expect("should get script");
-        assert_eq!(&script, "");
+    fn test_script_absent() -> Result<(), Error> {
+        let loc = ULoc::try_from("sr")?;
+        assert_eq!(loc.script(), None);
+        Ok(())
     }
 
     #[test]
-    fn test_country() {
-        let loc = ULoc::try_from("es-CO").expect("get es_CO locale");
-        let country = loc.country().expect("should get country");
-        assert_eq!(&country, "CO");
+    fn test_country() -> Result<(), Error> {
+        let loc = ULoc::try_from("es-CO")?;
+        assert_eq!(loc.country(), Some("CO".to_string()));
```

Comment:
nit: should we have this implement an Eq for `Option<&str>` too?

Reply:
Do you mean another test case for the no-country case? If so, Done.